### ISSUE TITLE
General 1st Interaction Feedback

### ISSUE THREAD

**Issue by metaskills** [lambci/cmda#1]
First, amazing tool and thanks so much for the work into this. I've been exploring using this tool for moving a key service to Lambda unlocked by EFS. So far so good when doing simple things like `ls` and `exec/echo` writes. But on my first upload I'm getting his error.

```
Invoking Lambda to copy files remotely...Rate Exceeded.
```

I can tell that the Lambda did compress a directory of files ~190MB and compressed ~80MB to the coda bucket. While this was happening, I was watching the CloudWatch Logs of the Lambda and this error above was reached long before I saw the cmda Lambda reach the "END RequestId: ..." part in the log. Eventually CloudWatch shoed this.

```
REPORT RequestId: f4fa646c-685c-44ab-8c2f-effc69a35cb3	Duration: 524145.03 ms	Billed Duration: 524200 ms	Memory Size: 3008 MB	Max Memory Used: 79 MB	
```

And after that I could confirm that nothing was uploaded. So thinking something happened during the decompress & write. FYI, I have verified that the cmda setup works by doing something like this.

```
$ cmda sh -c 'echo hello > /mnt/efs/someremotefile.txt'
```

Right now my EFS is in Bursting and General Purpose modes. Maybe the write to disk is slow and I should change it to provisioned? Other thoughts? Again, thanks for this heroic 🦸‍♂️🦸‍♀️ tool.

**Comment by mhart** [comment]
Interesting! I wonder where the Rate Exceeded is coming from.

Mine's in Bursting Mode too – and haven't run into this yet.

Can you copy files at all? Like even tiny ones?

**Comment by mhart** [comment]
I wonder if it's coming from Lambda? But that would suggest you're invoking the function *a lot*

https://aws.amazon.com/premiumsupport/knowledge-center/lambda-troubleshoot-throttling/

What's the output of `cmda info` btw?

**Comment by mhart** [comment]
Ah, interesting – I've encountered a scenario where it gets into a bit of a loop with large files and retries. I think it's related to this: https://github.com/aws/aws-sdk-js/issues/3311

Will play around a little more

**Comment by metaskills** [comment]
So I tried setting the concurrency to 5 and that just made things worse and slower with nothing happening. I took it back down to 1 again and tried a single file ~130KB and that's not going well now either. But I'll keep debugging with something small and go from there so I can report back.

**Comment by metaskills** [comment]
A few details about my setup that may be useful to know. I am using both the `AWS_PROFILE` and `CMDA_FUNCTION` environment variable. Maybe `CMDA_FUNCTION` env is not hooked up in all places?

**Comment by mhart** [comment]
Ok, try installing version v1.0.1 of the CLI and see how that goes (shouldn't need to update your stack or anything)

(`npm install -g cmda@latest` if you've installed it globally)

**Comment by mhart** [comment]
I suspect it's an issue to do with the function retrying the upload too many times if it gets corrupted. There's something up with the aws-sdk where it doesn't fully finish uploads in certain cases (which is why I opened [that issue](https://github.com/aws/aws-sdk-js/issues/3311)). That means that the tarball is corrupt, and cmda retries when it encounters such situations.

In my original testing it usually worked on the second or third upload, so I didn't have any max retries in there – I suspect you were getting throttled because it kept retrying over and over.

Have now added max retries, and also updated to my patched version of the aws sdk, so hopefully that fixes your issue. Otherwise it's something else and we need to keep exploring!

**Comment by metaskills** [comment]
I did notice when I downloaded the tarball from S3 to my Mac I could not decompress it. Trying 1.0.1 now.

**Comment by metaskills** [comment]
Updated and verified with cmda info. Still not working, my code is.

```shell
echo hello > hello.txt
cmda upload ./hello.txt /mnt/mydir/
```

I have verified that the file on S3 can be downloaded and uncompressed and is what I expect. Another aspect of my setup is that I am using the `lambci/lambda:build-nodejs12.x` container with shared volumes of `.:/var/task:delegated` and `~/.aws:/root/.aws:delegated` which I do often. So far uploading is the only issue. I'll try a download.

**Comment by mhart** [comment]
Wait, what? You're doing this in docker?

**Comment by mhart** [comment]
Oh I see, not running the Lambda function in docker – just using the build container as a sandbox to install and run the cmda cli?

**Comment by metaskills** [comment]
Correct! Keeps developers and CI/CD pipelines from having to use ad-hoc platform tools. BTW, I just verified that a download of a created text file via echo does not work. Never even makes the file on S3.

**Comment by mhart** [comment]
How did you create your cmda stack? And what version is it? `functionVersion` in `info`

**Comment by mhart** [comment]
Btw, when you say "still not working" do you mean you're still getting the same "Rate Exceeded" error?

**Comment by metaskills** [comment]
ClickOps via the app service catalog. `serverlessrepo:semanticVersion | 1.0.0` Guess I should tear that down and reset it up?

No, not working is just hanging here. `Invoking Lambda to copy files remotely..`

**Comment by mhart** [comment]
I want to hook up some sort of feedback mechanism, but it's tricky without creating some sort of websocket or something to communicate with – I think best strategy is tailing the logs and looking for a shared identifier, but would still be behind by a bit.

So just to reiterate where we're at – you're still unable to upload any files, even tiny ones?

**Comment by mhart** [comment]
You're saying this works:

```
cmda sh -c 'echo hello > /mnt/efs/someremotefile.txt'
cmda ls -l /mnt/efs/someremotefile.txt
```

But this doesn't:

```
echo hello > hello.txt
cmda upload hello.txt /mnt/efs/
cmda ls -l /mnt/efs/hello.txt
```

**Comment by mhart** [comment]
Btw, the `Invoking Lambda to copy files remotely` message can hang for quite a while if it's a number of files, especially if there are symlinks. EFS is slowwwww in such cases, and it just takes a while to write them

**Comment by metaskills** [comment]
Correct to the above on what works and what does not.

> Btw, the Invoking Lambda to copy files remotely message can hang for quite a while if it's a number of files

So this is just for the one `hello.txt` like the example above. I just deleted the cmda and am reinstalling it from the serverless app repo. Still looks like v1.0.0 on there. Is only the npm package 1.0.1?

**Comment by metaskills** [comment]
Sorry, not making progress, the simplest upload times out after 5 minutes. 

```shell
echo hello > hello.txt
cmda upload ./hello.txt /mnt/mydir/
```

```
Invoking Lambda to copy files remotely...Rate Exceeded.
```

**Comment by mhart** [comment]
Damn. Wish I could reproduce. No idea what that Rate Exceeded could be coming from.

Could it be a permissions thing maybe?

The files are being uploaded to the S3 bucket with `upload`, yeah? But you're saying they don't even make it to S3 with `download`?

I just uploaded 1.0.1 to SAR as well – just in case it makes a difference (can't see how it would for `upload`, but it might for `download`)

**Comment by metaskills** [comment]
> The files are being uploaded to the S3 bucket with upload, yeah?

Yup.

> But you're saying they don't even make it to S3 with download?

Correct.

I'm at a loss of things to try. I did fiddle with my security group some more. It did have ALL on since it was a dev test. Not sure what else to try. I know I don't want to use an EC2 instance and DataSync, so I might just write a quick Ruby Lambda that takes all files from an S3 bucket and syncs to EFS for this pipeline I'm doing. Would be willing to try other things if you had thoughts too.

**Comment by mhart** [comment]
Ah, it could be a VPC issue – ie, the Lambda can't access S3!

That would make sense I think because I've only tested from VPCs that have a NAT Gateway 🤦‍♂️

I'll see how to add a VPC endpoint for Lambda -> S3... hopefully the CFN isn't too painful

**Comment by mhart** [comment]
So I think you can do it like this:

```
aws ec2 create-vpc-endpoint --vpc-id vpc-1a2b3c4d --route-table-ids rtb-11aa22bb --service-name com.amazonaws.us-east-1.s3
```

I'll need to setup a new VPC to verify, but if you want to try that (just try using the main route table id for the VPC)

**Comment by mhart** [comment]
You can get the main route table like this:
```
aws ec2 describe-route-tables --filters Name=vpc-id,Values=vpc-1a2b3c4d Name=association.main,Values=true --query 'RouteTables[].RouteTableId'
```

**Comment by mhart** [comment]
Hmmm, maybe the main route table isn't the right one – maybe you have to use the route tables associated with the vpc subnets connected to the Lambda. God why is this stuff so complicated

**Comment by metaskills** [comment]
Thanks. I would not put too much stock into this. I think I can fix my VPC.

**Comment by mhart** [comment]
Still, would be nice to catch this and warn the user – and add documentation around it. At the least.

**Comment by metaskills** [comment]
Perhaps. I'm just bad at networking and using EFS just showed me that. My work initially used the public subnets which did not have a route to S3. I've since switched to private subnets and got things working. Good call out too on that corrupt stream data too. I certainly saw that at one time. 

Thanks again. I'll close this out now.

**Comment by mhart** [comment]
Cheers – I've published v1.0.2 that displays a nice error if it detects the Lambda can't communicate with S3.

Next step is looking to see if I can create a command that can automatically give your VPC access if you don't have it already.

**Comment by mhart** [comment]
Published v1.1.0 with a command to add a VPC endpoint (`cmda create-vpc-endpoint`) that should give the VPC access to the S3 bucket. This is also suggested if cmda detects that S3 isn't reachable by the Lambda function
